autoconnect/autoconnect-ws/autoconnect-ws-sm/src/identified/on_client_msg.rs

```diff
@@ -124,78 +127,161 @@ impl WebPushClient {
     }
 
-    async fn ack(&mut self, _updates: &[ClientAck]) -> Result<Vec<ServerMessage>, SMError> {
-        // TODO:
-        self.maybe_post_process_acks().await
+    /// Acknowledge receipt of one or more Push Notifications
+    async fn ack(&mut self, updates: &[ClientAck]) -> Result<Vec<ServerMessage>, SMError> {
+        trace!("WebPushClient:ack");
+        let _ = self.app_state.metrics.incr("ua.command.ack");
+
+        for notif in updates {
+            let pos = self
+                .ack_state
+                .unacked_direct_notifs
+                .iter()
+                .position(|n| n.channel_id == notif.channel_id && n.version == notif.version);
+            if let Some(pos) = pos {
+                debug!("Ack (Direct)";
+                       "channel_id" => notif.channel_id.as_hyphenated().to_string(),
+                       "version" => &notif.version
+                );
+                self.ack_state.unacked_direct_notifs.remove(pos);
+                self.stats.direct_acked += 1;
+                continue;
+            };
+
+            let pos = self
+                .ack_state
+                .unacked_stored_notifs
+                .iter()
+                .position(|n| n.channel_id == notif.channel_id && n.version == notif.version);
+            if let Some(pos) = pos {
+                debug!(
+                    "Ack (Stored)";
+                    "channel_id" => notif.channel_id.as_hyphenated().to_string(),
+                    "version" => &notif.version
+                );
+                let n = &self.ack_state.unacked_stored_notifs[pos];
+                // Topic/legacy messages have no sortkey_timestamp
+                if n.sortkey_timestamp.is_none() {
+                    debug!(
+                        "WebPushClient:ack removing Stored, sort_key: {}",
+                        &n.sort_key()
+                    );
+                    self.app_state
+                        .db
+                        .remove_message(&self.uaid, &n.sort_key())
+                        .await?;
+                }
+                self.ack_state.unacked_stored_notifs.remove(pos);
+                self.stats.stored_acked += 1;
+                continue;
+            };
+        }
+
+        if self.ack_state.unacked_notifs() {
+            // Wait for the Client to Ack all notifications before further
+            // processing
+            Ok(vec![])
+        } else {
+            self.post_process_all_acked().await
+        }
     }
 
-    async fn nack(&mut self, _code: Option<i32>) -> Result<(), SMError> {
-        unimplemented!();
+    /// Negative Acknowledgement (a Client error occurred) of one or more Push
+    /// Notifications
+    fn nack(&mut self, code: Option<i32>) {
+        trace!("WebPushClient:nack");
+        // only metric codes expected from the client (or 0)
+        let code = code
+            .and_then(|code| (301..=303).contains(&code).then_some(code))
+            .unwrap_or(0);
```

Comment:
I'd like to punt on this for now, it could be a separate issue (though it's not terribly important) just because I have no idea what the client might do if it begins trigger it.